python/cuml/common/logger.pyx

```diff
@@ -0,0 +1,288 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+
+from libcpp.string cimport string
+from libcpp cimport bool
+
+
+cdef extern from "cuml/common/logger.hpp" namespace "ML" nogil:
+    cdef cppclass Logger:
+        @staticmethod
+        Logger& get()
+        void setLevel(int level)
+        void setPattern(const string& pattern)
+        bool shouldLogFor(int level) const
+        string getPattern() const
+
+cdef extern from "cuml/common/logger.hpp" nogil:
+    void CUML_LOG_TRACE(const char* fmt, ...)
+    void CUML_LOG_DEBUG(const char* fmt, ...)
+    void CUML_LOG_INFO(const char* fmt, ...)
+    void CUML_LOG_WARN(const char* fmt, ...)
+    void CUML_LOG_ERROR(const char* fmt, ...)
+    void CUML_LOG_CRITICAL(const char* fmt, ...)
+
+    cdef int CUML_LEVEL_TRACE
+    cdef int CUML_LEVEL_DEBUG
+    cdef int CUML_LEVEL_INFO
+    cdef int CUML_LEVEL_WARN
+    cdef int CUML_LEVEL_ERROR
+    cdef int CUML_LEVEL_CRITICAL
+    cdef int CUML_LEVEL_OFF
+
+
+"""Enables all log messages upto and including `trace()`"""
+LEVEL_TRACE = CUML_LEVEL_TRACE
+
+"""Enables all log messages upto and including `debug()`"""
+LEVEL_DEBUG = CUML_LEVEL_DEBUG
+
+"""Enables all log messages upto and including `info()`"""
+LEVEL_INFO = CUML_LEVEL_INFO
+
+"""Enables all log messages upto and including `warn()`"""
+LEVEL_WARN = CUML_LEVEL_WARN
+
+"""Enables all log messages upto and include `error()`"""
+LEVEL_ERROR = CUML_LEVEL_ERROR
+
+"""Enables only `critical()` messages"""
+LEVEL_CRITICAL = CUML_LEVEL_CRITICAL
+
+"""Disables all log messages"""
+LEVEL_OFF = CUML_LEVEL_OFF
+
+
+def set_level(level):
```

Comment:
It would be neat to have a context `with` method besides this general set_level, i.e. so we can do:
```python 
with log_level(level):
    cuml_method()
```

Reply:
Ah... good point! Let me add it.

Reply:
@dantegd can you check now? I have added similar `with` statement support for both `set_level` and `set_pattern`.